### DISCUSSION TITLE
prefill html content inside `lexical/react` editor

### DISCUSSION THREAD

**Discussion by harish-sethuraman** [facebook/lexical#2320] [Q&A] [answered]
```js
import {$generateNodesFromDOM} from '@lexical/html';

// In the browser you can use the native DOMParser API to parse the HTML string.
const parser = new DOMParser();
const dom = parser.parseFromString(htmlString, textHtmlMimeType);

// In a headless environment you can use a package such as JSDom to parse the HTML string.
const dom = new JSDOM(htmlString);

// Once you have the DOM instance it's easy to generate LexicalNodes.
const nodes = $generateNodesFromDOM(editor, dom);

// Once you've generated LexicalNodes from your HTML you can now initialize an editor instance with the parsed nodes.
const editor = createEditor({ ...config, nodes });

// Or insert them at a selection.
const selection = $getSelection();
selection.insertNodes(nodes);
```
I see that we need the editor instance to `generateNodesfromDOM` how ever I'm using `lexical/react` and would Like to set the `initialEditorState` from HTML string. How do I do so? Can we somehow pre populate HTML string as editor content in `lexical/html`.

PS: Im using `LexicalRichTextPlugin`

**Comment by harish-sethuraman** [comment]
Hey bumping this in case it was missed

**Reply by trueadm** [reply]
Have you seen https://lexical.dev/docs/concepts/serialization#html---lexical?

**Reply by harish-sethuraman** [reply]
Yea I did. Here to get the editor what should we do? Should I get the editor from `useLexicalComposerContext` but when I do so it results in `Uncaught Error: LexicalComposerContext.useLexicalComposerContext: cannot find a LexicalComposerContext`

Also  `selection.insertNodes(nodes);` shouts the following
<img width="784" alt="Screenshot 2022-06-09 at 9 46 42 PM" src="https://user-images.githubusercontent.com/32865581/172895325-7e66e3dc-d837-4bfd-8a3a-00f9e8240257.png">
but the code example in the link you provided has them. I think I'm doing series of wrong steps here trying to implement this. Correct me if Im wrong thanks :)

**Reply by trueadm** [reply]
It seems like you're doing something wrong, but it's hard to be sure. You should be passing in a function to `<RichTextPlugin initialEditorState={fn}>`

That fn should be where you do this logic, as per the documentation. Right now, we don't pass `editor` to the function, but we can make a change so we do that? `insertNodes` definitely exists on `RangeSelection`

**Reply by trueadm** [reply]
https://github.com/facebook/lexical/pull/2378 adds passing `editor` to the function.

**Reply by harish-sethuraman** [reply]
Perfect thank you so much! I believe that should solve the issue.

**Reply by harish-sethuraman** [reply]
Now Im getting this error `Error: replace: cannot be called on root nodes` when I used the code provided in the doc. Made a sandbox to make things easier. 

https://codesandbox.io/embed/elated-keller-uolbew?fontsize=14&hidenavigation=1&theme=dark

**Reply by harish-sethuraman** [reply]
Any idea on this ? following up ;)

**Reply by harish-sethuraman** [reply]
I was getting `Error: replace: cannot be called on root nodes`

So I created a empty paragraph node and then selected it instead of selecting the root like given in the example.

```js
  function prepopulatedRichText(editor) {
    editor.update(() => {
      const parser = new DOMParser();
      const dom = parser.parseFromString(htmlString, textHtmlMimeType);
      const nodes = $generateNodesFromDOM(editor, dom);
      const emptypara = $createParagraphNode();
      emptypara.append($createTextNode(""));
      $getRoot().append(emptypara);
      emptypara.select();
      const selection = $getSelection();
      selection?.insertNodes(nodes);
    });
  }
```

**Reply by tejas-hosamani** [reply]
Hey @harish-sethuraman, your solution works. However, I can't create h1 elements inside paragraphs(it's against the standards, I believe.) 

But my question is, instead of creating a paragraph node, is it possible to create a 'div' node or something similar?

**Reply by tejas-hosamani** [reply]
Also, is it possible to add an object URL formatted image with this method?

**Comment by maltesa** [comment, marked as answer]
I wrote this Plugin that loads my legacy HTML Content. Maybe that helps:

```ts
import type { IRichText } from '@/components/RichTextEditor/types'
import { $generateNodesFromDOM } from '@lexical/html'
import { useLexicalComposerContext } from '@lexical/react/LexicalComposerContext'
import { $getRoot, CLEAR_HISTORY_COMMAND } from 'lexical'
import { useEffect } from 'react'

interface Props {
  state?: string | IRichText
}

export function RestoreEditorStatePlugin({ state }: Props) {
  const [editor] = useLexicalComposerContext()

  useEffect(() => {
    if (typeof state === 'string') {
      // Restore from HTML (Quilljs legacy)
      editor.update(() => {
        const parser = new DOMParser()
        const dom = parser.parseFromString(state, 'text/html')
        const nodes = $generateNodesFromDOM(editor, dom)

        const selection = $getRoot().select()
        selection.insertNodes(nodes)

        editor.dispatchCommand(CLEAR_HISTORY_COMMAND, undefined)
      })
    } else if (state?.editorState) {
      // Restore from SerializedEditorState
      const editorState = editor.parseEditorState(state.editorState)
      editor.setEditorState(editorState)
      editor.dispatchCommand(CLEAR_HISTORY_COMMAND, undefined)
    }
  }, [])

  return null
}
```

**Reply by tolkienfan2** [reply]
If this initializes your editor with your legacy HTML content, I don't understand how you can use 
`const [editor] = useLexicalComposerContext()`
 because when I try this, it returns the error 
`LexicalComposerContext.useLexicalComposerContext: cannot find a LexicalComposerContext`

**Reply by maltesa** [reply]
You have to use it within a LexicalComposer Component:

```typescript
interface Props {
  defaultValue?: any // string or editor state
}

const initialConfig = {
  namespace: 'MyEditor', 
  theme,
  onError,
};

export function RichTextEditor({ defaultValue }: Props) {
  return (
    <LexicalComposer initialConfig={initialConfig}>
      <RichTextPlugin
        contentEditable={<ContentEditable />}
        placeholder={<div>Enter text...</div>}
      />
      <RestoreEditorStatePlugin state={defaultValue} />
    </LexicalComposer>
  );
}

```

**Reply by tolkienfan2** [reply]
Thank you for this tip - I am now able to load some initial HTML.

**Reply by chalecki** [reply]
I've created pretty much the same plugin and I struggle with an extra empty `<p><br/></p>` 🤔. After reloading the editor, the content is not replaced but appended. Have you faced any issues like that?

**Reply by maltesa** [reply]
No, I haven't. I haven't done any updates to the project. Hence, it's still running on lexical 0.6 or something like that. I will try to remember to give an update as soon as I upgraded to the latest lexical version.

**Reply by chalecki** [reply]
Thank you @maltesa 🍺

**Reply by gajanan999** [reply]
Hi,

But what if I want to load the styles as well like coloring. it load the html tags properly but for styles?

**Reply by lwhiteley** [reply]
@gajanan999 you have to extend the TextNode as explained in the docs
https://lexical.dev/docs/concepts/serialization#handling-extended-html-styling

**Reply by lwhiteley** [reply]
@maltesa I have a similar solution but not all the same processes, i didnt clear the history but quick question for your solution.

Doesnt this violate the rules of hooks?
Semantically, would it make sense to use `useState`?

like so for eg.

```ts

export function RestoreEditorStatePlugin({ state }: Props) {
  const [editor] = useLexicalComposerContext()

  useState(() => {
    if (typeof state === 'string') {
      // Restore from HTML (Quilljs legacy)
      editor.update(() => {
        const parser = new DOMParser()
        const dom = parser.parseFromString(state, 'text/html')
        const nodes = $generateNodesFromDOM(editor, dom)

        const selection = $getRoot().select()
        selection.insertNodes(nodes)

        editor.dispatchCommand(CLEAR_HISTORY_COMMAND, undefined)
      })
    } 

   if (state?.editorState) {
      // Restore from SerializedEditorState
      const editorState = editor.parseEditorState(state.editorState)
      editor.setEditorState(editorState)
      editor.dispatchCommand(CLEAR_HISTORY_COMMAND, undefined)
    }
  })

  return null
}
```

What are your thoughts on that?

**Reply by maltesa** [reply]
Where do you see the rules of hooks violated?
Semantically, I think `useEffect` is correct. Restoring the editor state in a component is an Effect imo.

Is your solution working? To me, it looks like the function passed to `useState` is never called, is it?

**Reply by lwhiteley** [reply]
rules of hooks and `useEffect` stipulate that all reactive dependencies should be in the dependency array.

meaning, semantically, the dependency array of the above `useEffect` should be `[editor, state]`

`useState` accepts an initializer function or a value.. when useState uses an initializer function it is only called once when the component is mounted..(lazy initial state)

and yep the solution with useState works.

refs:
- https://react.dev/reference/react/useState#usestate

**Reply by maltesa** [reply]
Ah, I didn't know that you can pass a callback to useState :)
I thought rules of hooks just say that they must not be used conditionally and in loops.
Adding `editor` to the dependency array should not hurt. `state` would be better named `defaultState`. Hence, I wouldn't add it to the array to avoid overriding content changed by the user.

Concerning semantics, I like `useEffect` more because imo `useState` is for state and not for sideeffects. But I guess that's up to personal preference.

**Reply by lwhiteley** [reply]
I thought you had intentionally left the array empty, thats why i asked. If your goal is to only let it run once, then adding anything into the dependency array will run the effect again if any dependency reference changes. 
It gets even weirder in dev mode if the project uses StrictMode

i use the rules of hooks linter so the useEffect approach didnt work for me, hence why i ended up with `useState`

**Reply by maltesa** [reply]
yes, I wanted it to run once, that's why I left it empty. However, adding the editor to the array should not hurt, as I assume the editor reference will not change. And in case the reference changes you may want to reset the editor content.

You can disable the linter warning with `// eslint-disable-next-line react-hooks/exhaustive-deps`.

**Reply by lwhiteley** [reply]
No that's fine, I'll leave it as is. For me it makes more sense semantically to use the useState hook as I consider it just initializing the editor state for my use case and not really an effect that will run if anything changes.

**Reply by shammahk** [reply]
Thanks for the solution @maltesa, it helped.
I encountered an issue where nodes were duplicated, this was fixed by adding

```typescript
const root = $getRoot();
root.clear();
```
before inserting the nodes.

I hope this helps anyone who has such issues in the future.
